client/authority-discovery/src/worker.rs

```diff
@@ -137,12 +141,17 @@ where
 
 	/// Interval to be proactive, publishing own addresses.
 	publish_interval: Interval,
-	/// Interval on which to query for addresses of other authorities.
+	/// Interval at which to request addresses of authorities, refilling the pending lookups queue.
 	query_interval: Interval,
 	/// Interval on which to set the peerset priority group to a new random
 	/// set of addresses.
 	priority_group_set_interval: Interval,
 
+	/// Queue of throttled lookups pending to be passed to the network.
+	pending_lookups: Vec<AuthorityId>,
+	/// Set of in-flight lookups.
+	in_flight_lookups: HashMap<libp2p::kad::record::Key, AuthorityId>,
```

Comment:
You may be aware, but the way this field is used seems like using `HashMap<libp2p::kad::QueryId, AuthorityId>` instead, where `QueryId` is obtained from `libp2p-kad`, may offer some simplifications or improvements, namely:

  * No need to compute the `remote_key` in `handle_dht_value_found_event()`, passing a `QueryId` instead.
  * The still pending queries from the last run could be explicitly `finish()`ed (as in `libp2p::kad::QueryMut::finish()`) in `refill_pending_lookups_queue`. This way one could be certain about the upper bound on the still running queries. Of course, if the query interval is not shorter than the query timeout, that should also be guaranteed.

Unfortunately, due to the authority discovery being just a "client" of the `NetworkService` and thus not having direct access to the Kademlia APIs, this would require extending the API of `NetworkService` with appropriate asynchronous commands, so it may not be a simplification overall. Just thought I'd mention it, since it is possible to have more control over the queries, like aborting them, as far as `libp2p-kad` is concerned.

Reply:
> With the above in mind I would prefer not to do this pervasive change in this pull request. Instead I would open an issue to further discuss how we can improve the NetworkService Kademlia APIs.

Sounds good to me.

Reply:
Issue created here: https://github.com/paritytech/substrate/issues/7033